apps/prairielearn/src/pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.js

```diff
@@ -76,6 +76,14 @@ router.post(
           }
         },
       );
+    } else if (req.body.__action === 'update_group_roles') {
+      await groupAssessmentHelper.updateGroupRoles(
+        req.body,
```

Comment:
I'm not sure where exactly to put this comment, but: AFAICT, `updateGroupRoles` directly reads user/role IDs out of the request body and passes those directly to the SQL query, which in turn blindly inserts them into the database. This means that someone could in theory craft a malicious request that assigns arbitrary roles to arbitrary users. Of course, those users won't be in the group, so it shouldn't have any real negative impact. Still, it's best to always validate untrusted data coming from the client. In this case, you'd want to ensure group membership before assigning someone a role.

A related data issue: `group_user_roles` doesn't directly reference the `group_users` table, meaning that if a group user is deleted, their roles won't also be deleted. Not a huge issue, and I think this stems from the fact that `group_users` doesn't have an `id` PRIMARY KEY, so it's probably not super straightforward to fix.

Reply:
Gotcha. In our case, we check [here](https://github.com/PrairieLearn/PrairieLearn/blob/27c6835f7532c0c77d794603abbad71e43ead868/apps/prairielearn/src/lib/groups.js#L476-L481) that the user is in a group before doing any updates (using the assessment and user ID from `res.locals`). We use the user's own group ID when updating `group_user_roles`, so theoretically the user shouldn't be able to modify the roles of other groups.

I do see how a user could craft a request to add rows to `group_user_roles` with their own `group_id` but random non-member `user_id`s. We can add a check and corresponding test to make sure the `user_id`s coming in through the request exactly match those in the user's group.